### ISSUE TITLE
[sklearn#33554] chi2 feature selection is symmetric and cannot measure directional predictive power for classification

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#391]
**Mirrored from [scikit-learn#33554](https://github.com/scikit-learn/scikit-learn/issues/33554)**

Original title: chi2 feature selection is symmetric and cannot measure directional predictive power for classification
Original labels: New Feature, Needs Decision - Include Feature
Created: 2026-03-16T10:39:55Z

---

### Describe the issue

`sklearn.feature_selection.chi2` is widely used with `SelectKBest` for 
feature selection in classification. The documentation says it "weeds out 
the features that are the most likely to be independent of class and 
therefore irrelevant for classification."

The problem is that chi-square is a symmetric statistic. Mathematically:

$$\chi^2(X, Y) = \chi^2(Y, X)$$

The formula

$$\chi^2 = \sum_{i} \sum_{j} \frac{(O_{ij} - E_{ij})^2}{E_{ij}}$$

where $E_{ij} = \frac{R_i \times C_j}{N}$, is invariant to transposition 
of the contingency table. Swapping which variable is the row and which is 
the column does not change $O_{ij}$, $E_{ij}$, or the summation.

This is also visible in the source code:
```python
observed = safe_sparse_dot(Y.T, X)          # n_classes * n_features
feature_count = X.sum(axis=0).reshape(1, -1)
class_prob = Y.mean(axis=0).reshape(1, -1)
expected = np.dot(class_prob.T, feature_count)
return _chisquare(observed, expected)
```

Every operation here produces the same result if you swap feature and target.

### Why this matters for classification

Feature selection for classification is directional. We need to answer 
"does knowing this feature reduce uncertainty about the target." 
Chi-square cannot answer this. It answers "are these two variables 
associated" without any notion of direction.

### Concrete example

Consider a feature like zip code (200 unique values) and a binary target 
like churn (yes/no).

Chi-square can produce a high score even when zip code is useless for 
predicting churn. This happens because churn can partially distinguish 
between zip codes due to unequal group sizes across 200 categories. The 
score gets inflated by the target explaining the feature, not the feature 
explaining the target.

An asymmetric measure like Theil's U separates the two directions:

$$U(Y|X) = \frac{H(Y) - H(Y|X)}{H(Y)}$$

$$U(X|Y) = \frac{H(X) - H(X|Y)}{H(X)}$$

The numerator $H(Y) - H(Y|X)$ equals $H(X) - H(X|Y)$ — both equal 
mutual information $I(X;Y)$, which is symmetric just like chi-square. 
But the denominators $H(Y)$ and $H(X)$ are different, and that is what 
breaks the symmetry.

For a useless high cardinality feature:

- $U(Y|X) \approx 0$ — zip code tells us almost nothing about churn
- $U(X|Y) > 0$ — churn tells us something about zip code distribution
- $\chi^2$ gives one number for both directions with no way to separate them
```python
from dython.nominal import theils_u

# these two can give very different values
u_target_given_feature = theils_u(df['zipcode'], df['churn'])
u_feature_given_target = theils_u(df['churn'], df['zipcode'])

# chi2 gives the same score regardless of direction
from sklearn.feature_selection import chi2
chi2_scores, p_values = chi2(X, y)
```

### Current documentation is misleading

The docs use directional language like "irrelevant for classification" 
which implies the test measures predictive power toward the targ

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [5](https://github.com/godofecht/flow-scikit/issues/5) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
